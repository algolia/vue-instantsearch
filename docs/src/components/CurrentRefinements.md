---
title: CurrentRefinements
mainTitle: Components
layout: main.pug
category: Components
withHeadings: true
navWeight: 6
editable: true
githubSource: docs/src/components/current-refinements.md
---

Show the currently refined values and allow them to be unset.

<a class="btn btn-static-theme" href="stories/?selectedKind=CurrentRefinements">ðŸ•¹ try out live</a>

## Usage

```html
<ais-current-refinements :clears-query="true"></ais-current-refinements>
```

## Props

Name | Type | Default | Description | Required
---|---|---|---|---
transformItems | Function | | Allows you to format and change the attributes | -
clearsQuery | Boolean | `false` | Should the 'clear all' button also clear the query? | -
excludedAttributes | Array | `[]` | Attributes not to show or clear | -
transformItems | `(items: object[]) => object[]` | `x => x` | Function which receives the items, which will be called before displaying them. Should return a new array with the same shape as the original array. Useful for mapping over the items to transform, remove or reorder them | -
classNames | Object | | Override class names | no

## Slots

name | scope | Description
---|---|---
`default` | `{ items: Item[], refine: Item => void, clearAll: () => void }` | Override how all the items look
`item`Â | `{ item: Item, refine: Item => void }` | Override how an item looks
`clearAllLabel`Â | `{ items: Item[] }` | Override how the "clear all" button looks

## CSS classes

Here's a list of CSS classes exposed by this widget. To better understand the underlying
DOM structure, have a look at the generated DOM in your browser.

Note that you can pass the prop `class-names`, with an object of class names and their replacement to override this.

Class name | Description
---|---
`ais-CurrentRefinements` | Container class
`ais-CurrentRefinements--noRefinement` | Modifier on the container when no refinements are enabled.
`ais-CurrentRefinements-item` | An item
`ais-CurrentRefinements-delete` | The button in the item to delete that refinement
`ais-CurrentRefinements-item--selected` | Selected item
