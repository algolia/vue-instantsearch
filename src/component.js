import suit from './suit.js';

export default {
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  created() {
    this.widgetFactory = this.connector(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdate = Object.keys(widgetParams).some(name =>
        widgetParams[name](previous[name], next[name])
      );

      if (shouldUpdate) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  methods: {
    suit(...args) {
      return suit(this.widgetName, ...args);
    },
    updateData(state = {}) {
      this.state = state;
    },
  },
};
