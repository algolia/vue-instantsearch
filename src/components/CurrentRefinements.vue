<template>
  <div :class="suit('')" v-if="state">
    <slot v-bind="state">
      <ul :class="suit('list')">
        <li v-for="item in state.refinements" :key="item.attributeName">
          <span :class="suit('item')">
            <span :class="suit('label')">{{item.attributeName | capitalize}}: {{item.computedLabel}}</span>
            <button :class="suit('delete')" @click="state.refine(item)">âœ•</button>
          </span>
        </li>
      </ul>
      <button :class="suit('reset')" @click="state.clearAllClick()" v-if="state.refinements.length > 0">
        Clear all
      </button>
    </slot>
  </div>
</template>

<script>
import algoliaComponent from '../component';
import { connectCurrentRefinedValues } from 'instantsearch.js/es/connectors';

export default {
  mixins: [algoliaComponent],
  props: {
    attributes: {
      type: Array,
      required: false,
    },
    clearsQuery: {
      type: Boolean,
      default: false,
      required: false,
    },
    excludedAttributes: {
      type: Array,
      required: false,
      validator(val) {
        // TODO: make this work with the connector (IS.js doesn't expose it)
        if (Array.isArray(val) && val.length > 0) {
          console.warn(
            '`excludedAttributes` is not implemented on CurrentRefinements'
          );
          return false;
        }
        return true;
      },
    },
  },
  data() {
    return {
      widgetName: 'CurrentRefinements',
    };
  },
  beforeCreate() {
    this.connector = connectCurrentRefinedValues;
  },
  computed: {
    widgetParams() {
      return {
        attributes: this.attributes,
        clearsQuery: this.clearsQuery,
      };
    },
  },
  filters: {
    capitalize(value) {
      if (!value) return '';
      return (
        value
          .toString()
          .charAt(0)
          .toUpperCase() + value.toString().slice(1)
      );
    },
  },
};
</script>
