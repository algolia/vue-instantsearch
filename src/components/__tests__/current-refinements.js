import { mount } from '@vue/test-utils';
import CurrentRefinements from '../CurrentRefinements.vue';
import { __setState } from '../../component';
jest.mock('../../component');

it('renders correctly (empty)', () => {
  __setState({
    refinements: [],
  });
  const wrapper = mount(CurrentRefinements);
  expect(wrapper.html()).toMatchSnapshot();
});

it('renders correctly (with refinements)', () => {
  __setState({
    refinements: [
      {
        attributeName: 'brands',
        computedLabel: 'apple',
      },
      {
        attributeName: 'colors',
        computedLabel: 'red',
      },
      {
        attributeName: 'requirements',
        computedLabel: 'free',
      },
    ],
  });
  const wrapper = mount(CurrentRefinements);
  expect(wrapper.html()).toMatchSnapshot();
});

it('calls `refine` with an item', () => {
  __setState({
    refinements: [
      {
        attributeName: 'brands',
        computedLabel: 'apple',
      },
      {
        attributeName: 'colors',
        computedLabel: 'red',
      },
      {
        attributeName: 'requirements',
        computedLabel: 'free',
      },
    ],
    refine: jest.fn(),
  });
  const wrapper = mount(CurrentRefinements);
  wrapper.find('.ais-CurrentRefinements-delete').trigger('click');

  expect(wrapper.vm.state.refine).toHaveBeenLastCalledWith({
    attributeName: 'brands',
    computedLabel: 'apple',
  });
});

it('calls `refine` with clear all', () => {
  __setState({
    refinements: [
      {
        attributeName: 'brands',
        computedLabel: 'apple',
      },
      {
        attributeName: 'colors',
        computedLabel: 'red',
      },
      {
        attributeName: 'requirements',
        computedLabel: 'free',
      },
    ],
    clearAllClick: jest.fn(),
  });
  const wrapper = mount(CurrentRefinements);
  wrapper.find('.ais-CurrentRefinements-reset').trigger('click');

  expect(wrapper.vm.state.clearAllClick).toHaveBeenCalled();
});
