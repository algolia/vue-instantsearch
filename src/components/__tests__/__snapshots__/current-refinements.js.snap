// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly (empty) 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
  </ul>
</div>

`;

exports[`renders correctly (with refinements) 1`] = `

<div class="ais-CurrentRefinements">
  <ul class="ais-CurrentRefinements-list">
    <li>
      <span class="ais-CurrentRefinements-item">
        <span class="ais-CurrentRefinements-label">
          Brands: apple
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
    <li>
      <span class="ais-CurrentRefinements-item">
        <span class="ais-CurrentRefinements-label">
          Colors: red
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
    <li>
      <span class="ais-CurrentRefinements-item">
        <span class="ais-CurrentRefinements-label">
          Requirements: free
        </span>
        <button class="ais-CurrentRefinements-delete">
          ✕
        </button>
      </span>
    </li>
  </ul>
  <button class="ais-CurrentRefinements-reset">
    Clear all
  </button>
</div>

`;
